sklearn/impute.py

```diff
@@ -1246,15 +1258,14 @@ def transform(self, X):
 
         imputer_mask, features = self._get_missing_features_info(X)
 
-        if self.features == "missing-only":
+        if self.features in ("missing-only", "not-constant"):
             features_diff_fit_trans = np.setdiff1d(features, self.features_)
             if (self.error_on_new and features_diff_fit_trans.size > 0):
                 raise ValueError("The features {} have missing values "
                                  "in transform but have no missing values "
                                  "in fit.".format(features_diff_fit_trans))
 
-            if (self.features_.size > 0 and
-                    self.features_.size < self._n_features):
```

Comment:
I removed the first condition. If we want only features with missing, and there's not any, then the mask should be empty. Before, it would return the mask of all features.